gobblin-core/src/main/java/gobblin/writer/http/AbstractHttpWriter.java

```diff
@@ -97,27 +110,107 @@ public long bytesWritten() throws IOException {
   }
 
   /**
+   * Send and process the request. If it's a retry request, skip onNewRecord method call and go straight sending request.
    * {@inheritDoc}
    */
   @Override
   public void writeImpl(D record) throws IOException {
-    this.curRequest = onNewRecord(record, this.curRequest);
-    if (this.curRequest.isPresent()) {
-      ListenableFuture<HttpResponse> responseFuture = sendRequest(this.curRequest.get());
-      waitForResponse(responseFuture);
-      try {
-        processResponse(responseFuture.get());
-      } catch (InterruptedException | ExecutionException e) {
-        throw new RuntimeException(e);
+    if (!isRetry()) {
+      //If currentRequest is still here, it means this is retry request.
+      //In this case, don't invoke onNewRecord again as onNewRecord is not guaranteed to be idempotent.
+      //(e.g: If you do batch processing duplicate record can go in, etc.)
+      curRequest = onNewRecord(record);
+    }
+
+    if (curRequest.isPresent()) {
+      ListenableFuture<CloseableHttpResponse> responseFuture = sendRequest(curRequest.get());
+      try (CloseableHttpResponse response = waitForResponse(responseFuture)) {
+        processResponse(response);
+      }
+      curRequest = Optional.absent(); //Clear request if successful
+    }
+    numRecordsWritten++;
```

Comment:
Synced offline. numRecordsWritten is only increased when it's successfully written, thus no retry would trigger numRecordsWritten++.